hbase-server/src/main/java/org/apache/hadoop/hbase/master/region/MasterRegion.java

```diff
@@ -143,13 +148,23 @@ private void shutdownWAL() {
     }
   }
 
+  @RestrictedApi(explanation = "Should only be called in tests", link = "",
+      allowedOnPath = ".*/src/test/.*")
+  public void setTestFailure() {
+    this.updateFailForTest = true;
+  }
+
   public void update(UpdateMasterRegion action) throws IOException {
     try {
+      if (updateFailForTest) {
+        // test for HBASE-29251
+        throw new IOException("Update failed");
+      }
```

Comment:
Not a good way to test his, but since MasterRegion is final class, extending it is also not possible.
Though we can reproduce the exact issue with the test.

Reply:
I think it should be fine to remove `final` from MasterRegion class so that we can extend it for testing purpose.

Reply:
MasterRegion has a private constructor, that's why we mark it as final.

Since UpdateMasterRegion is just a interface, I think it is very easy to verify the changes?

Reply:
The problem with HRegion is that, we don't want to have custom HRegion that doesn't work for all cases. We want to inject error for batchMutate() only when region moves.

Reply:
Any custom implementation of UpdateMasterRegion also needs custom hooks in Procedure executor classes, which make it more complicated to test.

Reply:
You need to define what you want to test, if you just want to make sure that if there is an exception you will call abort, it is very easy, and even do not need to bring up a cluster.

If you want to do something like a integration tests, you can extend HRegion, and there are bunch of ways to decide whether to throw an exception in batchMutate method. You can make the specific HRegion implementation an inner class of the testcase, and set a static field in the test class to control whether to throw an exception...